### PR TITLE
[feat] moe: initial implementation of Top2Gating

### PR DESCRIPTION
**Pull request by msbaines**
# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [ ] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

## What does this PR do?
Fixes # (issue).

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: fairscale/nn/moe/top2gate.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+            size of model embedding dimension
+        num_experts (ints):
+            number of experts in model
+    """
+
+    wg: torch.nn.Linear
+
+    def __init__(self, model_dim: int, num_experts: int,) -> None:
+        super().__init__()
+        self.wg = torch.nn.Linear(num_experts, model_dim, bias=False)
```
Comment (ngoyal2707) [inline fairscale/nn/moe/top2gate.py:107]: nit: extract this to separate module? don't need to make it configurable for now though

Comment (msbaines) [reply fairscale/nn/moe/top2gate.py:107]: so:

def __init__(self, wg: torch.nn.Linear) -> None:

File: fairscale/nn/moe/top2gate.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+# Code is inspired by Top2GatingOnLogits from lingvo:
+#   https://github.com/tensorflow/lingvo/blob/21b8106c5f1d30a196c98eedc441d4fd70833b11/lingvo/core/moe_layers.py#L477
+
+from typing import Tuple
+
+import torch
+from torch import Tensor
+import torch.nn.functional as F
+
+gumble = torch.distributions.gumbel.Gumbel(0, 1)  # type: ignore
```
Comment (TimDettmers) [inline fairscale/nn/moe/top2gate.py:16]: `gumble` -> `gumbel`

**Review by TimDettmers — APPROVED** [review]
Looks great! An implementation with masks is so much cleaner than with indexing. I was skeptical about the inputs into the Gumbel max trick, I have seen some formulations that required unnormalized log probabilities (https://arxiv.org/abs/1611.01144), but I did a quick test and this implementation is correct and will yield the right distribution.

File: fairscale/nn/__init__.py [hunk 3]
```diff
@@ -1,3 +1,4 @@
+from .moe import Top2Gate
```
Comment (min-xu-ai) [inline fairscale/nn/__init__.py:1]: add copyright?

File: tests/nn/moe/test_top2gating.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    assert torch.equal(combine_weights.bool(), dispatch_mask)
+    assert torch.all(torch.sum(dispatch_mask, axis=(1, 3)) <= capacity)
+    assert torch.all(combine_weights >= 0.0)
+    assert torch.all(combine_weights <= 1.0)
+    weights_sum = torch.sum(combine_weights).item()
+    assert round(weights_sum) == pytest.approx(weights_sum)
+    assert weights_sum == pytest.approx(36.0)
+
+
+def test_top1s():
```
Comment (min-xu-ai) [inline tests/nn/moe/test_top2gating.py:35]: add a reference to the paper where this test is based on?

File: tests/nn/moe/test_top2gating.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    assert pytest.approx(l_aux.item(), 0.0283)
+    assert combine_weights.shape == (3, 12, 6, 4)
+    assert dispatch_mask.shape == (3, 12, 6, 4)
+    assert torch.equal(combine_weights.bool(), dispatch_mask)
+    assert torch.all(torch.sum(dispatch_mask, axis=(1, 3)) <= capacity)
+    assert torch.all(combine_weights >= 0.0)
+    assert torch.all(combine_weights <= 1.0)
+    weights_sum = torch.sum(combine_weights).item()
+    assert round(weights_sum) == pytest.approx(weights_sum)
+    assert weights_sum == pytest.approx(36.0)
```
Comment (min-xu-ai) [inline tests/nn/moe/test_top2gating.py:32]: where is 36 from?

**Review by min-xu-ai — APPROVED** [review]
Looks awesome. I am still checking the top2gating code sequence. But looks good already!
